### ISSUE TITLE
UseXRay(...) causes concurrency exception

### ISSUE THREAD

**Issue by yreynhout** [aws/aws-xray-sdk-dotnet#120]
In my ASP NET Core 2.1 web app's `Configure(IApplicationBuilder app)` I set up XRay using

```
app.UseXRay(new DynamicSegmentNamingStrategy("My.Server"), this.configuration);
```

Now, I have a test suite that runs against this web app (using TestServer) and on occasion I get the following exception:

```
[Critical] Application startup exception : Microsoft.AspNetCore.Hosting.Internal.WebHost: 
      InvalidOperationException:
      Message    : 'Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
      StackTrace : '
            at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
            at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
            at Amazon.XRay.Recorder.Core.Internal.Emitters.JsonSegmentMarshaller..ctor()
            at Amazon.XRay.Recorder.Core.AWSXRayRecorder.InitializeInstance(IConfiguration configuration)
            at Amazon.XRay.Recorder.Handlers.AspNetCore.Internal.AWSXRayMiddleware..ctor(RequestDelegate next, SegmentNamingStrategy segmentNamingStrategy, IConfiguration configuration)
         --- End of stack trace from previous location where exception was thrown ---
            at Microsoft.Extensions.Internal.ActivatorUtilities.ConstructorMatcher.CreateInstance(IServiceProvider provider)
            at Microsoft.Extensions.Internal.ActivatorUtilities.CreateInstance(IServiceProvider provider, Type instanceType, Object[] parameters)
            at Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.<>c__DisplayClass4_0.<UseMiddleware>b__0(RequestDelegate next)
            at Microsoft.AspNetCore.Builder.Internal.ApplicationBuilder.Build()
            at Microsoft.AspNetCore.Hosting.Internal.WebHost.BuildApplication()
      TargetSite : 'Void ThrowInvalidOperationException_ConcurrentOperationsNotSupported()
```

It looks like `JsonMapper` is using plain dictionaries (indirectly called by `AWSXRayRecorder -> JsonSegmentMarshaller ->`), which are not protected from being accessed concurrently, and thus causes the standard .NET dictionary exception. Am I right in assuming that AWSXRayRecorder should only be initialized once per process space and that I should pick an overload that takes the AWSXRayRecorder instance? Obviously, that puts the burden on me (I'm fine with that). I just wanted to raise awareness others might run into this as well.

I have to wonder if this is somehow connected the `LazyDefaultRecorder` in `AWSXRayRecorderImpl` and how it gets initialized.

**Comment by srprash** [comment]
Hi @yreynhout 
You are right that the AWSXRayRecorder gets initialized only once per process and not per request. We use the singleton instance for the recorder every time. 
For confirming this being related to the LazyDefaultRecorder, I'll have to dig a but deeper.

**Comment by srprash** [comment]
Hi @yreynhout 
I was testing this for threading issues but couldn't reproduce it. I tested it with a console app which creates segments for multiple threads, and then also tried a webapp invoked by multiple threads concurrently. I would like to know more about how you're testing your webapp. If you can provide a sample code that would be helpful.
Looking at the stack trace, my guess for your case is that if your application or test is initializing the recorder concurrently, it might be hitting [this code](https://github.com/aws/aws-xray-sdk-dotnet/blob/40d3c0e924eb8e41c1b28fab073cfd36b4c986f3/sdk/src/Core/ThirdParty/LitJson/JsonMapper.cs#L134) in `JsonMapper` where we initialize a bunch of dictionaries like `base_exporters_table`, etc.
I can see a potential problem with these dictionaries based on how the initialization is done in an application. Let me know if you can provide more insights from your use case.

**Comment by yreynhout** [comment]
I'll try to whip up an example by the end of this week and post a gist here (I'm no longer active on the project that ran into this but I still would like to get it resolved).

**Comment by srprash** [comment]
Hi @yreynhout 
I'm closing out this issue due to inactivity. Feel free to reopen and provide more info on the issue if you can.
Thanks!

**Comment by yreynhout** [comment]
Sorry about that - life got in the way :-) I'm fine with closing - I'm sure that _if_ it resurfaces someone will open another issue.
